### ISSUE TITLE
TensorRT8.0 MSA op can't be aligned with nn.MultiheadAttention when passing key_padding_mask parameter

### ISSUE THREAD

**Issue by lippman1125** [NVIDIA/TensorRT#1483]
## Description

I use TRT to build a MultiheadAttention network,but Its result is different from nn.MultiheadAttention.If I set key_padding_mask=None in nn.MultiheadAttention and has_mask=False in TRT, their results are equal.
So I suppose that something wrong with mask parameter of TRT.  

Run command: 
```
python3 attention_layer_mask_test.py --seq 6 --mask 6 --half
```
The code as follow,:
```
import argparse
import ctypes
import json
import numpy as np
import os
import os.path
import re
import sys
import time

# CUDA
import pycuda.driver as cuda
import pycuda.autoinit
# TensorRT
import tensorrt as trt
import torch
import torch.nn as nn


def set_tensor_name(tensor, prefix, name):
    tensor.name = prefix + name

def set_output_name(layer, prefix, name, out_idx = 0):
    set_tensor_name(layer.get_output(out_idx), prefix, name)

def parse_args():
    parser = argparse.ArgumentParser(description='attention test')
    parser.add_argument('--seq', type=int, default=256, help='Sequence')
    parser.add_argument('--half', dest='half', action='store_true', help="network precision")
    parser.add_argument('--mask', type=int, default=128, help="mask length")
    return parser.parse_args()

"""
TensorRT Initialization
"""
TRT_LOGGER = trt.Logger(trt.Logger.INFO)
trt_version = [int(n) for n in trt.__version__.split('.')]
handle = ctypes.CDLL("libnvinfer_plugin.so", mode=ctypes.RTLD_GLOBAL)
if not handle:
    raise RuntimeError("Could not load plugin library. Is `libnvinfer_plugin.so` on your LD_LIBRARY_PATH?")

def cal_err_pointwise(a, b):
    a = a.astype(np.float)
    b = b.astype(np.float)
    if np.all(np.equal(a, b)):
        return 0.0
    diff = a - b
    denom = np.abs(b) + np.ones_like(b) * 1e-10
    ratio = np.abs(diff) / denom
    err_mean = np.mean(ratio)
    err_max = np.max(ratio)
    return err_mean, err_max, np.max(np.abs(diff))

if __name__ == "__main__":
    args = parse_args()
    print(args)
    
    seq = args.seq
    mask = args.mask
    FLOAT16_TYPE = args.half

    # init library
    trt.init_libnvinfer_plugins(TRT_LOGGER, "")
    # get plugin
    plg_registry = trt.get_plugin_registry()
    qkv2_plg_creator = plg_registry.get_plugin_creator("CustomQKVToContextPluginDynamic", "1", "")

    # init param
    workspace_size = 1000
    explicit_batch_flag = 1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)

    # build engine
    builder = trt.Builder(TRT_LOGGER)
    network = builder.create_network(explicit_batch_flag)
    builder_config = builder.create_builder_config()
    builder_config.max_workspace_size = workspace_size * (1024 * 1024)
    TRT_TYPE = trt.float32
    if FLOAT16_TYPE is True:
        print("build fp16")
        builder_config.set_flag(trt.BuilderFlag.FP16)
        builder_config.set_flag(trt.BuilderFlag.STRICT_TYPES)
        TRT_TYPE = trt.float16

    if trt_version[0] >= 8:
        tactic_source = 1 << int(trt.TacticSource.CUBLAS) | 1 << int(trt.TacticSource.CUBLAS_LT)
        builder_config.set_tactic_sources(tactic_source)

    # build network
    hidden_size = 768
    num_heads = 4
    has_mask = False
    N = num_heads
    H = hidden_size//num_heads
    Wall = np.random.randn(3*hidden_size*hidden_size)
    Ball = np.random.randn(3*hidden_size)
    Wqkv = trt.Weights(np.ascontiguousarray(np.ascontiguousarray(Wall.reshape((3, N, H, N, H)).transpose((1, 0, 2, 3, 4))), dtype=np.float32))
    Bqkv = trt.Weights(np.ascontiguousarray(np.ascontiguousarray(Ball.reshape((3, N, H)).transpose((1, 0, 2))), dtype=np.float32))

    Wout = np.random.randn(hidden_size*hidden_size)
    Bout = np.random.randn(hidden_size)
    W_aout = trt.Weights(np.ascontiguousarray(Wout, dtype=np.float32).flatten())
    B_aout = trt.Weights(np.ascontiguousarray(Bout, dtype=np.float32).flatten())

    # config network
    inputs = network.add_input(name="inputs", dtype=TRT_TYPE, shape=(seq,1,hidden_size,1,1))
    imask = network.add_input(name="imask", dtype=trt.float32, shape=(1,1))
    assert (len(inputs.shape) == 5)
    S, B, hidden_size, _, _ = inputs.shape

    mult_all = network.add_fully_connected(inputs, 3 * hidden_size, Wqkv, Bqkv)
    set_output_name(mult_all, "attention/", "qkv_mult")
    print("mult_all output shape={}, dtype={}".format(mult_all.get_output(0).shape, mult_all.get_output(0).dtype))
    # QKV2CTX
    pf_type = trt.PluginField("type_id", np.array([int(TRT_TYPE)], np.int32), trt.PluginFieldType.INT32)
    pf_hidden_size = trt.PluginField("hidden_size", np.array([hidden_size], np.int32), trt.PluginFieldType.INT32)
    pf_num_heads = trt.PluginField("num_heads", np.array([num_heads], np.int32), trt.PluginFieldType.INT32)
    pf_has_mask = trt.PluginField("has_mask", np.array([int(has_mask)], np.int32), trt.PluginFieldType.INT32)
    pfc = trt.PluginFieldCollection([pf_hidden_size, pf_num_heads, pf_has_mask, pf_type])
    qkv2ctx_plug = qkv2_plg_creator.create_plugin("qkv2ctx", pfc)
    qkv_in = [mult_all.get_output(0)]
    if has_mask:
        qkv_in.append(imask)
    qkv2ctx = network.add_plugin_v2(qkv_in, qkv2ctx_plug)
    set_output_name(qkv2ctx, "attention/", "context_layer")

    attention_heads = qkv2ctx.get_output(0)
    attention_out_fc = network.add_fully_connected(attention_heads, hidden_size, W_aout, B_aout)
    attention_out = attention_out_fc.get_output(0)
    set_tensor_name(attention_out, "attention/", "multi_head_out")

    network.mark_output(attention_out)
    network.get_output(0).dtype=TRT_TYPE
    print("network output shape={} dtype={}".format(network.get_output(0).shape, network.get_output(0).dtype))

    engine = builder.build_engine(network, builder_config)

    TRT_LOGGER.log(TRT_LOGGER.VERBOSE, "Serializing Engine...")
    serialized_engine = engine.serialize()
    engine_file = "attentionlayer_half.trt" if FLOAT16_TYPE else "attentionlayer_float.trt"
    TRT_LOGGER.log(TRT_LOGGER.INFO, "Saving Engine to {:}".format(engine_file))
    with open(engine_file, "wb") as fout:
        fout.write(serialized_engine)
    TRT_LOGGER.log(TRT_LOGGER.INFO, "Done.")

    # Inference
    with open("{}".format(engine_file), "rb") as f, \
            trt.Runtime(TRT_LOGGER) as runtime, \
            runtime.deserialize_cuda_engine(f.read()) as engine, \
            engine.create_execution_context() as context:

        if FLOAT16_TYPE is True:
            INPUT_DATA_TYPE = np.float16
        else:
            INPUT_DATA_TYPE = np.float32
        stream = cuda.Stream()
        print("bind0 shape={}".format(engine.get_binding_shape(0)))
        print("bind1 shape={}".format(engine.get_binding_shape(1)))
        print("bind2 shape={}".format(engine.get_binding_shape(2)))
        host_in1 = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(0)), dtype=INPUT_DATA_TYPE)
        host_in2 = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(1)), dtype=INPUT_DATA_TYPE)
        host_out = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(2)), dtype=INPUT_DATA_TYPE)
        devide_in1 = cuda.mem_alloc(host_in1.nbytes)
        devide_in2 = cuda.mem_alloc(host_in2.nbytes)
        devide_out = cuda.mem_alloc(host_out.nbytes)
        # bind device mem
        bindings = [int(devide_in1), int(devide_in2), int(devide_out)]

        # infer
        emb = np.random.randn(*engine.get_binding_shape(0)).astype(INPUT_DATA_TYPE)
        mask_arr = np.array([[mask]]).astype(np.float32)
        np.copyto(host_in1, emb.ravel())
        np.copyto(host_in2, mask_arr.ravel())
        # print("host_in={}".format(host_in))
        cuda.memcpy_htod_async(devide_in1, host_in1, stream)
        cuda.memcpy_htod_async(devide_in2, host_in2, stream)
        context.execute_async(bindings=bindings, stream_handle=stream.handle)
        cuda.memcpy_dtoh_async(host_out, devide_out, stream)
        stream.synchronize()
        nret = np.array(host_out).reshape(seq, 1, hidden_size)
        print(nret)
        # nret = nret[:mask]

        MAt = nn.MultiheadAttention(hidden_size, num_heads)
        MAt.in_proj_weight = nn.Parameter(torch.from_numpy(Wall.reshape(3*hidden_size, hidden_size)), requires_grad=False)
        MAt.in_proj_bias = nn.Parameter(torch.from_numpy(Ball), requires_grad=False)

        MAt.out_proj.weight = nn.Parameter(torch.from_numpy(Wout.reshape(hidden_size, hidden_size)), requires_grad=False)
        MAt.out_proj.bias = nn.Parameter(torch.from_numpy(Bout), requires_grad=False)

        # prepare mask
        np_mask = np.zeros((1,seq))
        np_mask[0,mask:] = 1
        np_mask = torch.from_numpy(np_mask)
        np_mask = np_mask.to("cuda")
        print(np_mask)
        # prepare embedding
        emb = torch.from_numpy(emb.squeeze(axis=(3,4)))
        emb = emb.to("cuda")
        MAt.half()
        MAt.to("cuda")
        mret, att_w = MAt(emb, emb, emb, key_padding_mask = None)
        mret = mret.cpu().numpy()
        print(mret)
        # mret = mret[:mask]

        err_mean, err_max, _ =cal_err_pointwise(nret.flatten(), mret.flatten())
        print("err mean = {}, err_max = {}".format(err_mean, err_max))


```

## Environment

**TensorRT Version**: 8.0.1.6
**NVIDIA GPU**: T4
**NVIDIA Driver Version**: 450.51.06
**CUDA Version**: 11.0
**CUDNN Version**: 8.0.3
**Operating System**: centos7
**Python Version (if applicable)**: 3.6.8
**Tensorflow Version (if applicable)**: 2.6.0
**PyTorch Version (if applicable)**: 1.9.0
**Baremetal or Container (if so, version)**:

**Comment by ttyio** [comment]
Hello @lippman1125 ,

Just checked the code, this plugin is not designed in a clean way, the maskPtr in this plugin is declared as fp32, but read as int32 internally. see

https://github.com/NVIDIA/TensorRT/blob/master/plugin/bertQKVToContextPlugin/qkvToContext.cu#L257

Could you try change your code to use int32 mask? thanks

         mask_arr = np.array([[int(mask)]]).astype(np.int32)

**Comment by ttyio** [comment]
@lippman1125 , hmmm, sorry more updates to the code:

we need allocate the maskptr using `float32`  (the underhood is int32 actually)

       host_in2 = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(1)), dtype=np.float32)

 then we cannot use `np.copyto`, because we are copying int32 data to float32 storage (maybe you can try different `casting` to see if any of them works), my suggesting is to use `cuda.register_host_memory` to avoid `np.copyto`, the code looks like:

      host_in1 = cuda.register_host_memory(np.ascontiguousarray(emb.ravel()))
      host_in2 = cuda.register_host_memory(np.ascontiguousarray(mask_arr.ravel()))

**Comment by lippman1125** [comment]
> @lippman1125 , hmmm, sorry more updates to the code:
> 
> the maskptr using `float32` (the underhood is int32 actually)
> 
> ```
>    host_in2 = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(1)), dtype=np.float32)
> ```
> 
> then we cannot use `np.copyto`, because we are copying int32 data to float32 storage (maybe you can try different `casting` to see if any of them works), my suggesting is to use `cuda.register_host_memory` to avoid `np.copyto`, the code looks like:
> 
> ```
>   host_in1 = cuda.register_host_memory(np.ascontiguousarray(emb.ravel()))
>   host_in2 = cuda.register_host_memory(np.ascontiguousarray(mask_arr.ravel()))
> ```

I modify the code as follow, but it is not effective.
```
-        host_in2 = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(1)), dtype=INPUT_DATA_TYPE)
+        host_in2 = cuda.pagelocked_empty(trt.volume(engine.get_binding_shape(1)), dtype=np.int32)

-        mask_arr = np.array([[mask]]).astype(np.float32)
+        mask_arr = np.array([[mask]]).astype(np.int32)
         np.copyto(host_in1, np.ascontiguousarray(emb.ravel()))
         np.copyto(host_in2, np.ascontiguousarray(mask_arr.ravel()))
```

**Comment by ttyio** [comment]
Hi @lippman1125 ,
Here is a code modified based on your original source, this code is to prove the trt plugin result with mask and without mask are the same, so we need pass the same `--seq` and `--mask` .
[run.zip](https://github.com/NVIDIA/TensorRT/files/7335363/run.zip)

**Comment by ttyio** [comment]
@lippman1125 ,  I use the mask == seq test, and compare the result to prove the mask implementation in the plugin is correct. 
There is no limitation that we have to use mask == seq in real use case.

**Comment by ttyio** [comment]
close since no activity for more than 3 weeks, please reopen if you still have question, thanks!

**Comment by linchaozhang** [comment]
@lippman1125 
did you solve this problem？ nn.MultiheadAttention convert to TRT
